shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Generator/Utxo.hs

```diff
@@ -104,17 +105,40 @@ import Test.Shelley.Spec.Ledger.Generator.Core
     findStakeScriptFromCred,
     genNatural,
   )
+import Test.Shelley.Spec.Ledger.Generator.MetaData (genMetaData)
 import Test.Shelley.Spec.Ledger.Generator.Trace.DCert (genDCerts)
 import Test.Shelley.Spec.Ledger.Generator.Update (genUpdate)
 
--- | Generates a transaction in the context of the LEDGER STS environment and state.
+-- | Generates a transaction in the context of the LEDGER STS environment
+-- and state.
+--
+--  A generated transaction may not have sufficient spending balance and
+-- need to be discarded. In this case we retry the generator,
+-- up to 'genTxRetries' times, before failing hard with an error.
+--
+-- Note: the spending balance emerges from inputs, refund withdrawals,
+-- certificate deposits and fees (which in turn depend on number of
+-- inputs, outputs, witnesses, metadata etc.). It's hard to avoid this
+-- completely, but in practice it is relatively easy to calibrate
+-- the generator 'Constants' so that there is sufficient spending balance.
 genTx ::
   (HasCallStack, Mock c) =>
   GenEnv c ->
   LedgerEnv ->
   (UTxOState c, DPState c) ->
   Gen (Tx c)
-genTx
+genTx ge@(GenEnv _ (Constants {genTxRetries})) =
```

Comment:
@nfrisby I've baked "retrying" into `genTx` - you could use `genTxRetries` in your consensus tests - or set it to 1 if you want to keep the retry mechanism on consensus side

Reply:
That will simplify my use-case nicely. Thanks!